test/e2e/user_configured_cabundle_test.go

```diff
@@ -0,0 +1,692 @@
+		serverCABundleConfigMapName = types.NamespacedName{
+			Name:      "test-server-ca-bundle",
+			Namespace: testServerNamespace,
+		}
+
+		serviceName1      = "hello-world-1"
+		svcTLSSecretName1 = types.NamespacedName{
+			Name:      serviceName1,
+			Namespace: testServerNamespace,
+		}
+		svcDeploymentName1 = types.NamespacedName{
+			Name:      serviceName1,
+			Namespace: testServerNamespace,
+		}
+		svcLabels1 = map[string]string{
+			"app": serviceName1,
+		}
+		svcPort1 = int32(8082)
+		svcName1 = types.NamespacedName{
+			Name:      serviceName1,
+			Namespace: testServerNamespace,
+		}
+		routeName1 = types.NamespacedName{
+			Name:      serviceName1,
+			Namespace: testServerNamespace,
+		}
+
+		serviceName2      = "hello-world-2"
+		svcTLSSecretName2 = types.NamespacedName{
+			Name:      serviceName2,
+			Namespace: testServerNamespace,
+		}
+		svcDeploymentName2 = types.NamespacedName{
+			Name:      serviceName2,
+			Namespace: testServerNamespace,
+		}
+		svcLabels2 = map[string]string{
+			"app": serviceName2,
+		}
+		svcPort2 = int32(8083)
+		svcName2 = types.NamespacedName{
+			Name:      serviceName2,
+			Namespace: testServerNamespace,
+		}
+		routeName2 = types.NamespacedName{
+			Name:      serviceName2,
+			Namespace: testServerNamespace,
+		}
+	)
+
+	svcCACert1, svcCAKey1, err := generateCAKeyPair("operator-e2e-1")
+	if err != nil {
+		t.Fatalf("failed to generate first server CA certificate: %v", err)
+	}
+	svcCert1, svcKey1, err := generateServerCertificate(svcCACert1, svcCAKey1, serviceName1, []string{serviceName1, fmt.Sprintf("%s.%s.svc", serviceName1, testServerNamespace)})
+	if err != nil {
+		t.Fatalf("failed to generate first server certificate: %v", err)
+	}
+	clientCert, clientKey, err := generateClientCertificate(svcCACert1, svcCAKey1, "test-client")
+	if err != nil {
+		t.Fatalf("failed to generate client certificate: %v", err)
+	}
+	svcCACert2, svcCAKey2, err := generateCAKeyPair("operator-e2e-2")
+	if err != nil {
+		t.Fatalf("failed to generate first server CA certificate: %v", err)
+	}
+	svcCert2, svcKey2, err := generateServerCertificate(svcCACert2, svcCAKey2, serviceName2, []string{serviceName2, fmt.Sprintf("%s.%s.svc", serviceName2, testServerNamespace)})
+	if err != nil {
+		t.Fatalf("failed to generate second server certificate: %v", err)
+	}
+
+	// Create the configmap with the custom CA certificates which will be
+	// used for verifying services' certificates.
+	adminCABundleConfigMapName := controller.AdminCAConfigMapName()
+	adminCABundleConfigMap := buildConfigMap(adminCABundleConfigMapName,
+		map[string]string{
+			"ca-bundle.crt": encodeCert(svcCACert1),
+		})
+	if err := kclient.Create(ctx, adminCABundleConfigMap); err != nil {
+		t.Fatalf("failed to create configmap %q: %v", adminCABundleConfigMapName, err)
+	}
+	defer func() {
+		if err := kclient.Delete(ctx, adminCABundleConfigMap); err != nil {
+			t.Fatalf("failed to delete configmap %q: %v", adminCABundleConfigMapName, err)
+		}
+	}()
```

Comment:
We've been moving towards using `t.Cleanup`:
```suggestion
	t.Cleanup(func() {
		if err := kclient.Delete(ctx, adminCABundleConfigMap); err != nil {
			t.Fatalf("failed to delete configmap %q: %v", adminCABundleConfigMapName, err)
		}
	})
```
Most test code still uses `defer` even when it could use `t.Cleanup`, so I don't mind if you prefer to continue with `defer` here.  We'll need to overhaul the test code in the future anyway to change old uses of `defer` to `t.Cleanup` where we can.

Reply:
suggestion incorporated.